submissions/krylenger/a-tiny-JS-world_post_OOP/index.js

```diff
@@ -0,0 +1,140 @@
+/* Refer to https://github.com/OleksiyRudenko/a-tiny-JS-world for the task details */
+
+const dog = {
+    type: "animal",
+    legs: 4,
+    hands: 0,
+    gender: "male",
+    friends: ["Anna", "Eugene"],
+  },
+  cat = {
+    type: "animal",
+    legs: 4,
+    hands: 0,
+    gender: "female",
+    friends: ["Anna", "Eugene"],
+  },
+  woman = {
+    type: "human",
+    legs: 2,
+    hands: 2,
+    gender: "female",
+    friends: ["Eugene", "Murzik", "Sharik"],
+  },
+  man = {
+    type: "human",
+    legs: 2,
+    hands: 2,
+    gender: "male",
+    friends: ["Anna", "Murzik", "Sharik"],
+  },
+  catWoman = {
+    type: "cat-woman",
+    legs: 2,
+    hands: 2,
+    gender: "female",
+    friends: ["Anna", "Murzik"],
+  };
+
+const introduceSelf = ({
+  type,
+  name,
+  gender,
+  legs,
+  hands,
+  friends,
+}) => `Hi! My name is <em>${name}</em> and I'm <em>${type}</em>. 
+    My gender is <em>${gender}</em>. I have <em>${legs}</em> legs and <em>${hands}</em> hands. 
+    These are my friends: <em>${friends.join(", ")}</em>.<em> 
+    </em><br><br>`;
+
+const inhabitantSays = (self) => ({
+  introduction: () => {
+    return introduceSelf(self);
+  },
+});
+
+const createInhabitant = ({ type, legs, hands, gender, friends }) => {
+  let self = {
+    type,
+    legs,
+    hands,
+    gender,
+    friends,
+  };
+  return Object.assign(self, inhabitantSays(self));
+};
+
+const catSays = (self) => ({
+  introduction: () => {
+    return `Meow! Meow meow <em>${self.name}</em> meow meow <em>${
+      self.type
+    }</em>. 
+      meow meow <em>${self.gender}</em>. meow <em>${self.legs}</em> meow <em>${
+      self.hands
+    }</em> meow. 
+      Meow meow meow: <em>${self.friends.join(", ")}</em>.<em> 
+      </em><br><br>`;
+  },
+});
+
+const dogSays = (self) => ({
+  introduction: () => {
+    return `Woof! Woof woof <em>${self.name}</em> woof woof <em>${
+      self.type
+    }</em>. 
+      woof woof <em>${self.gender}</em>. woof <em>${self.legs}</em> woof <em>${
+      self.hands
+    }</em> woof. 
+      Woof woof woof: <em>${self.friends.join(", ")}</em>.<em> 
+      </em><br><br>`;
+  },
+});
+
```

Comment:
Quite a creative idea to have distinct communications for humans, dogs, and cats.
There is a pattern in `catSays` and `dogSays`. How about having a single yet customizable method for animals? This way adding more species of animals cheaper.

In other words how would you change your code in this section if you were asked to introduce other species of animals?

Reply:
✅ Done
How do you find this implementation?
I mean the part of modifying the carnivoraPetSays function for catWoman. 
My concern that the part with `if if (self.species === 'catWoman')` doesn't look very nice but I need to use the same introduction method for catWoman and she doesn't have paws.

Reply:
@krylenger well, yes, looks not really nice. The problem is that it is an all-in-one function versus [SRP](https://en.wikipedia.org/wiki/Single-responsibility_principle). Split it.

First, split out methods to assign sayings to an inhabitants of this type (animals ans animal-like).

Then have a method for output creation. Obviously CatWoman should employ a method for humans in these regards.

The beauty of compositional design pattern is that one can combine small or even tiny chunks of properties and methods in order to achieve the goals yet keeping code simple and concise.

Those chunks should stay as abstract (agnostic of a specific species in your case) as possible. Whenever a composable method needs to know what entity class method or property it will be attributed to it means that the implementation is poor.

Do you want to take this challenge? We might get finished with what's already in place.

Reply:
I really appreciate your feedback!
Certainly I want to take this challenge to understand the key principles better.
After your comments I refactored my classes and methods on them to follow composition design pattern. 
Now comparing to the previous version I can clearly see the weak parts of my previous code.
What about now? How do you find it? Are there some more work to be done to improve this code? Thanks in advance.